File: scripts/convert_dict.js
/**
 * 词典转换脚本
 * 将 kajweb/dict 项目的 JSON 格式转换为 Vignette 应用需要的 TXT 格式
 */

const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

// 法语字母替换表
const FR_EN = [
  ['é', 'e'], ['ê', 'e'], ['è', 'e'], ['ë', 'e'],
  ['à', 'a'], ['â', 'a'], ['ç', 'c'],
  ['î', 'i'], ['ï', 'i'],
  ['ô', 'o'],
  ['ù', 'u'], ['û', 'u'], ['ü', 'u'],
  ['ÿ', 'y']
];

function replaceFrenchChars(text) {
  if (!text) return '';
  let result = text;
  for (const [fr, en] of FR_EN) {
    result = result.split(fr).join(en);
  }
  return result;
}

function extractPhonetic(wordData) {
  const content = wordData?.content?.word?.content || {};
  const usphone = content.usphone || '';
  const ukphone = content.ukphone || '';
  const phonetic = usphone || ukphone;
  return replaceFrenchChars(phonetic).trim();
}

function extractMeanings(wordData) {
  const content = wordData?.content?.word?.content || {};
  const trans = content.trans || [];
  
  const meanings = [];
  for (const t of trans) {
    let pos = (t.pos || '').trim();
    let tranCn = replaceFrenchChars((t.tranCn || '').trim());
    
    if (pos && tranCn) {
      if (!pos.endsWith('.')) {
        pos = pos + '.';
      }
      meanings.push(`${pos} ${tranCn}`);
    } else if (tranCn) {
      meanings.push(tranCn);
    }
  }
  
  return meanings;
}

function processZipFile(zipPath, allWords) {
  const AdmZip = require('adm-zip');
  const zip = new AdmZip(zipPath);
  const entries = zip.getEntries();
  
  let count = 0;
  
  for (const entry of entries) {
    if (!entry.entryName.endsWith('.json')) continue;
    
    const content = entry.getData().toString('utf8');
    const lines = content.split('\n');
    
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      
      let wordData;
      try {
        wordData = JSON.parse(trimmed);
      } catch (e) {
        continue;
      }
      
      let headWord = (wordData.headWord || '').trim().toLowerCase();
      if (!headWord) continue;
      
      headWord = replaceFrenchChars(headWord);
      const phonetic = extractPhonetic(wordData);
      const meanings = extractMeanings(wordData);
      
      if (!allWords.has(headWord)) {
        allWords.set(headWord, { phonetic, meanings });
      } else {
        const existing = allWords.get(headWord);
        if (!existing.phonetic && phonetic) {
          existing.phonetic = phonetic;
        }
        const existingSet = new Set(existing.meanings);
        for (const m of meanings) {
          if (!existingSet.has(m)) {
            existing.meanings.push(m);
            existingSet.add(m);
          }
        }
      }
      
      count++;
    }
  }
  
  return count;
}

function writeOutput(allWords, outputPath) {
  const sorted = [...allWords.entries()].sort((a, b) => a[0].localeCompare(b[0]));
  
  const lines = [];
  for (const [word, data] of sorted) {
    const { phonetic, meanings } = data;
    
    let line = word;
    if (phonetic) {
      line = `${word} [${phonetic}]`;
    }
    
    if (meanings.length > 0) {
      line = line + ' ' + meanings.join(' ');
    }
    
    lines.push(line);
  }
  
  fs.writeFileSync(outputPath, lines.join('\n'), 'utf8');
  console.log(`写入 ${sorted.length} 个单词到 ${outputPath}`);
}

async function main() {
  const dictDir = 'D:\\Downloads\\dict-master\\dict-master';
  const bookDir = path.join(dictDir, 'book');
  const outputPath = 'D:\\DevEcoStudioProjects\\Vignette\\entry\\src\\main\\resources\\rawfile\\wordbooks\\master.txt';
  
  const allWords = new Map();
  let totalCount = 0;
  
  // 检查 adm-zip 是否安装
  try {
    require.resolve('adm-zip');
  } catch (e) {
    console.log('安装 adm-zip...');
    require('child_process').execSync('npm install adm-zip', { cwd: __dirname, stdio: 'inherit' });
  }
  
  if (fs.existsSync(bookDir)) {
    const files = fs.readdirSync(bookDir);
    for (const filename of files) {
      if (!filename.endsWith('.zip')) continue;
      
      const zipPath = path.join(bookDir, filename);
      console.log(`处理: ${filename}`);
      const count = processZipFile(zipPath, allWords);
      totalCount += count;
      console.log(`  -> ${count} 个单词`);
    }
  }
  
  console.log(`\n总共处理: ${totalCount} 条记录`);
  console.log(`去重后: ${allWords.size} 个唯一单词`);
  
  writeOutput(allWords, outputPath);
  
  let withPhonetic = 0;
  let withMeanings = 0;
  for (const [, data] of allWords) {
    if (data.phonetic) withPhonetic++;
    if (data.meanings.length > 0) withMeanings++;
  }
  
  console.log('\n统计:');
  console.log(`  有音标: ${withPhonetic} (${(withPhonetic * 100 / allWords.size).toFixed(1)}%)`);
  console.log(`  有释义: ${withMeanings} (${(withMeanings * 100 / allWords.size).toFixed(1)}%)`);
}

main().catch(console.error);

File: scripts/convert_dict.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
词典转换脚本
将 kajweb/dict 项目的 JSON 格式转换为 Vignette 应用需要的 TXT 格式
"""

import os
import json
import zipfile
from typing import Dict, List

# 法语字母替换表
FR_EN = [
    ['é', 'e'], ['ê', 'e'], ['è', 'e'], ['ë', 'e'],
    ['à', 'a'], ['â', 'a'], ['ç', 'c'],
    ['î', 'i'], ['ï', 'i'],
    ['ô', 'o'],
    ['ù', 'u'], ['û', 'u'], ['ü', 'u'],
    ['ÿ', 'y']
]

def replace_french_chars(text: str) -> str:
    """替换法语字母为英语字母"""
    for fr, en in FR_EN:
        text = text.replace(fr, en)
    return text

def extract_phonetic(word_data: dict) -> str:
    """提取音标，优先美音"""
    content = word_data.get('content', {}).get('word', {}).get('content', {})
    usphone = content.get('usphone', '')
    ukphone = content.get('ukphone', '')
    
    phonetic = usphone if usphone else ukphone
    phonetic = replace_french_chars(phonetic)
    return phonetic.strip()

def extract_meanings(word_data: dict) -> List[str]:
    """提取释义"""
    content = word_data.get('content', {}).get('word', {}).get('content', {})
    trans = content.get('trans', [])
    
    meanings = []
    for t in trans:
        pos = t.get('pos', '').strip()
        tran_cn = t.get('tranCn', '').strip()
        tran_cn = replace_french_chars(tran_cn)
        
        if pos and tran_cn:
            if not pos.endswith('.'):
                pos = pos + '.'
            meanings.append(f"{pos} {tran_cn}")
        elif tran_cn:
            meanings.append(tran_cn)
    
    return meanings

def process_zip_file(zip_path: str, all_words: Dict[str, dict]) -> int:
    """处理ZIP文件"""
    count = 0
    
    with zipfile.ZipFile(zip_path, 'r') as zf:
        for name in zf.namelist():
            if name.endswith('.json'):
                with zf.open(name) as f:
                    content = f.read().decode('utf-8')
                    for line in content.strip().split('\n'):
                        line = line.strip()
                        if not line:
                            continue
                        
                        try:
                            word_data = json.loads(line)
                        except json.JSONDecodeError:
                            continue
                        
                        head_word = word_data.get('headWord', '').strip().lower()
                        if not head_word:
                            continue
                        
                        head_word = replace_french_chars(head_word)
                        
                        phonetic = extract_phonetic(word_data)
                        meanings = extract_meanings(word_data)
                        
                        if head_word not in all_words:
                            all_words[head_word] = {
                                'phonetic': phonetic,
                                'meanings': meanings
                            }
                        else:
                            if not all_words[head_word]['phonetic'] and phonetic:
                                all_words[head_word]['phonetic'] = phonetic
                            existing_meanings = set(all_words[head_word]['meanings'])
                            for m in meanings:
                                if m not in existing_meanings:
                                    all_words[head_word]['meanings'].append(m)
                                    existing_meanings.add(m)
                        
                        count += 1
    
    return count

def write_output(all_words: Dict[str, dict], output_path: str):
    """写入输出文件"""
    sorted_words = sorted(all_words.items(), key=lambda x: x[0])
    
    with open(output_path, 'w', encoding='utf-8') as f:
        for word, data in sorted_words:
            phonetic = data['phonetic']
            meanings = data['meanings']
            
            if phonetic:
                line = f"{word} [{phonetic}]"
            else:
                line = word
            
            if meanings:
                line = line + " " + " ".join(meanings)
            
            f.write(line + '\n')
    
    print(f"写入 {len(sorted_words)} 个单词到 {output_path}")

def main():
    dict_dir = r"D:\Downloads\dict-master\dict-master"
    book_dir = os.path.join(dict_dir, "book")
    output_path = r"D:\DevEcoStudioProjects\Vignette\entry\src\main\resources\rawfile\wordbooks\master.txt"
    
    all_words: Dict[str, dict] = {}
    total_count = 0
    
    if os.path.exists(book_dir):
        for filename in os.listdir(book_dir):
            if filename.endswith('.zip'):
                zip_path = os.path.join(book_dir, filename)
                print(f"处理: {filename}")
                count = process_zip_file(zip_path, all_words)
                total_count += count
                print(f"  -> {count} 个单词")
    
    print(f"\n总共处理: {total_count} 条记录")
    print(f"去重后: {len(all_words)} 个唯一单词")
    
    write_output(all_words, output_path)
    
    with_phonetic = sum(1 for w in all_words.values() if w['phonetic'])
    with_meanings = sum(1 for w in all_words.values() if w['meanings'])
    print(f"\n统计:")
    print(f"  有音标: {with_phonetic} ({with_phonetic*100/len(all_words):.1f}%)")
    print(f"  有释义: {with_meanings} ({with_meanings*100/len(all_words):.1f}%)")

if __name__ == '__main__':
    main()
